pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,309 @@
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+// The internal sync interval of kubelet will drive the refresh of data.
+func (m *metadataVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+func (m *metadataVolume) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a metadata volume %v for pod %v at %v", m.volName, m.pod.UID, dir)
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := m.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, m.pod, *m.opts, m.mounter)
+	if err != nil {
+		return err
+	}
+	if err := wrapped.SetUpAt(dir); err != nil {
+		return err
+	}
+	var changed bool
+	// files are fully regenerated only if requested metadata changed
+	if changed, err = m.isDataChanged(); err == nil && changed {
+		err = m.writeData()
+	}
+	return err
+}
+
+// isDataChanged iterate over all the entries to check wether at least one
+// file needs to be updated.
+func (m *metadataVolume) isDataChanged() (bool, error) {
+	errlist := []error{}
+	for fieldReference, fileName := range m.fieldReferenceFileNames {
+		var data string
+		var err error
+		if data, err = fieldpath.ExtractFieldPathAsString(m.pod, fieldReference); err != nil {
+			glog.Error(err)
+			errlist = append(errlist, err)
+		}
+		if isFileToGenerate(path.Join(m.GetPath(), fileName), sortLines(data)) {
+			return true, utilErrors.NewAggregate(errlist)
+		}
+	}
+	return false, utilErrors.NewAggregate(errlist)
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+const currentTmpDir = ".current_tmp"
+
+// writeData writes requested metadata in specified files.
+//
+// The file visible in this volume are symlinks to files in the '.current'
+// directory. Actual files are stored in an hidden timestamped directory which is
+// symlinked to by '.current'. The timestamped directory and '.current' symlink
+// are created in the plugin root dir.  This scheme allows the files to be
+// atomically updated by changing the target of the '.current' symlink.  When new
+// data is available:
+//
+// 1.  A new timestamped dir is created by writeDataInTmpDir
+// 2.  Symlinks for new files are created if needed
+// 3.  The previous timestamped directory is detected reading the '.current' symlink
+// 4.  In case no symlink exists then it's created and func returns (first execution)
+// 5.  In case symlink exists a new temporary symlink is created .current_tmp
+// 6.  .current_tmp is renamed to .current
+// 7.  The previous timestamped directory is removed
+func (m *metadataVolume) writeData() error {
+	var tempDir string
```

Comment:
@sdminonne I think it would enhance this code if we changed 'tmpDir' to 'timestampDir' throughout.